vim25/types/json_test.go

```diff
@@ -0,0 +1,648 @@
+				FileOperation: VirtualDeviceConfigSpecFileOperationCreate,
+				Device: &VirtualVmxnet3{
+					VirtualVmxnet: VirtualVmxnet{
+						VirtualEthernetCard: VirtualEthernetCard{
+							VirtualDevice: VirtualDevice{
+								Key: 3,
+							},
+							MacAddress: "00:11:22:33:44:55:66:88",
+						},
+					},
+				},
+			},
+		},
+	}); err != nil {
+		t.Fatal(err)
+	}
+	act, exp := w.String(), virtualMachineConfigSpecWithDeviceChangesJSON
+	if act != exp {
+		t.Errorf("act json != exp json\nact=%s\nexp=%s", act, exp)
+	}
+}
+
+func TestJSONUnmarshalVirtualMachineConfigSpec(t *testing.T) {
+	dec := json.NewDecoder(strings.NewReader(virtualMachineConfigSpecWithVAppConfigJSON))
+	dec.SetDiscriminator("_typeName", "_value", "", json.DiscriminatorToTypeFunc(TypeFunc()))
+
+	var obj VirtualMachineConfigSpec
+	if err := dec.Decode(&obj); err != nil {
+		t.Fatal(err)
+	}
+
+	var w bytes.Buffer
+	enc := json.NewEncoder(&w)
+	enc.SetIndent("", "  ")
+	enc.SetDiscriminator("_typeName", "_value", "")
+
+	if err := enc.Encode(obj); err != nil {
+		t.Fatal(err)
+	}
+
+	act, exp := w.String(), virtualMachineConfigSpecWithVAppConfigJSON
+	if act != exp {
+		t.Errorf("act json != exp json\nact=%s\nexp=%s", act, exp)
+	}
+}
+
+func TestJSONUnmarshalVirtualMachineConfigInfo(t *testing.T) {
+	f, err := os.Open("./testdata/virtualMachineConfigInfo.json")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer f.Close()
+
+	dec := json.NewDecoder(f)
+	dec.SetDiscriminator("_typeName", "_value", "", json.DiscriminatorToTypeFunc(TypeFunc()))
+
+	var obj VirtualMachineConfigInfo
+	if err := dec.Decode(&obj); err != nil {
+		t.Fatal(err)
+	}
+
+	if diff := cmp.Diff(obj, virtualMachineConfigInfoObjForTestData); diff != "" {
+		t.Errorf("mismatched VirtualMachineConfigInfo: %s", diff)
+		fmt.Println(diff)
+	}
+}
+
+const virtualMachineConfigSpecWithDeviceChangesJSON = `{
+  "_typeName": "VirtualMachineConfigSpec",
+  "name": "Hello, world.",
+  "deviceChange": [
+    {
+      "_typeName": "VirtualDeviceConfigSpec",
+      "operation": "add",
+      "fileOperation": "create",
+      "device": {
+        "_typeName": "VirtualVmxnet3",
+        "key": 3,
+        "macAddress": "00:11:22:33:44:55:66:88"
+      }
+    }
+  ]
+}
+`
+
+const virtualMachineConfigSpecWithVAppConfigJSON = `{
+  "_typeName": "VirtualMachineConfigSpec",
+  "name": "Hello, world.",
+  "vAppConfig": {
+    "_typeName": "VmConfigSpec",
+    "product": [
+      {
+        "_typeName": "VAppProductSpec",
+        "operation": "add",
+        "info": {
+          "_typeName": "VAppProductInfo",
+          "key": 1,
+          "name": "p1"
+        }
+      }
+    ],
+    "installBootRequired": false
+  }
+}
+`
+
+func mustParseTime(layout, value string) time.Time {
+	t, err := time.Parse(layout, value)
+	if err != nil {
+		panic(err)
+	}
+	return t
+}
+
+func addrOfMustParseTime(layout, value string) *time.Time {
+	t := mustParseTime(layout, value)
+	return &t
+}
+
+func addrOfBool(v bool) *bool {
```

Comment:
Fine to keep these, but note we also have [types.New{Bool,Int32,Int64}](https://github.com/vmware/govmomi/blob/73bed2d89ecb73aa3968fda483d35bb4cb2a333c/vim25/types/helpers.go#L26-L36)